### ISSUE TITLE
react-redux typing doesnt inherit static members for connected components

### ISSUE THREAD

**Issue by pdeva** [DefinitelyTyped/DefinitelyTyped#20985]
- [x] I tried using the `@types/react-redux` package and had problems.
`Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @tkqubo @thasner @kenzierocks @clayne11 @tansongyang @NicholasBoll @mDibyo

if you do:

```ts
const MyComp_Connected = connect(mapStateToProps)(MyComp)

//then call
MyComp_Connected.staticMethod1(); //Typescript will complain at compile time 
//of not seeing this method
//however the code will work at runtime
```

TypeScript complains about static member not being present on `connect()-ed` class, even though it works at runtime.

The typings need to be updated to pass through static members to connected components

**Comment by mDibyo** [comment]
@pdeva You wanna take a shot at solving this one?

**Comment by pdeva** [comment]
i would love to but my typescript-foo is not strong enough. cant even figure out most of the current definition...

**Comment by Kovensky** [comment]
The typings are actually correct. See the source code at https://github.com/reactjs/react-redux/blob/master/src/components/connectAdvanced.js#L115

The component that `react-redux` generates does _not_ inherit from your component, but from `React.Component`. It [_renders_ your component](https://github.com/reactjs/react-redux/blob/master/src/components/connectAdvanced.js#L249-L258), and passes `props` through, but does _not_ inherit or otherwise do anything else actually related to your component.

tl;dr It is not possible to access either static or instance props of a component that has been wrapped by `connect`.

---

However, ideally, it should be possible to call `getWrappedInstance()` on a `ref` to the `connect`ed component and get something of the original class' type. That currently is not possible. But if you _actually_ need to do that something probably went wrong with your encapsulation.

**Comment by pdeva** [comment]
>EDIT: nevermind, it will do a copy of the static members on the resulting class as the last thing it does before returning the wrapped class 😕

so we are back to the original issue of the typings needing update

**Comment by kenzierocks** [comment]
I'll take a look tonight.

**Comment by Kovensky** [comment]
@pdeva it's still not as simple; I'm not sure if it can even be expressed in typescript. *Only* the static members are copied -- instance members are not accessible unless you call `getWrappedInstance()`.

**Comment by pdeva** [comment]
@Kovensky if it cant be expressed in TypeScript, then a relevant issue should be opened on TypeScript repo to point out this use case.

**Comment by adyry** [comment]
Ok, so with most of props->state lifecycle methods being deprecated, how we deal with **static getDerivedStateFromProps()** ?

**Comment by SugarDarius** [comment]
I had to use //@ts-ignore to make static getDerivedStateFromProps() works ...

**Comment by dan-kez** [comment]
Just want to chime in on this - we're migrating to Typescript and this is a particular painpoint

**Comment by MrKou47** [comment]
Does anyone has hack way for solve this error?

**Comment by jamesreggio** [comment]
I've proposed a fix for this problem in #33446. (My branch also includes updates to `hoist-non-react-statics`, which I need to land in #33445 before this will work.)

If anybody would like to test this out, I'd be much obliged.
